### ISSUE TITLE
Can't rebuild winax with electron 12.0.2

### ISSUE THREAD

**Issue by chpuget** [durs/node-activex#92]
Hi

I am using winax with node and it's working well !
I am now testing with electron, using the latest which is : 12.0.2 (npm i -D electron@latest)
{
  node: '14.16.0',
  v8: '8.9.255.20-electron.0',
  uv: '1.40.0',
  zlib: '1.2.11',
  brotli: '1.0.9',
  ares: '1.16.1',
  modules: '87',
  nghttp2: '1.41.0',
  napi: '7',
  llhttp: '2.1.3',
  openssl: '1.1.1',
  icu: '68.1',
  unicode: '13.0',
  electron: '12.0.2',
  chrome: '89.0.4389.90'
}

I have VS2017 and VS2019, as well as nodeJS 14.16.0 on my machine

Normal installation with node : (works well)
>  npm install winax --msvs_version=2019 (using 2017 or 2019)

Rebuilding for node, following your instructions in https://www.npmjs.com/package/winax 
> npm rebuild winax --runtime=electron --target=12.0.2 --dist-url=https://electronjs.org/headers --build-from-source
--> see errors below

Can you please help to understand the issue here ?

Thanks and regards

Chris

> winax@3.1.1 install C:\<dev folder>\electronPOC\node_modules\winax
> node-gyp rebuild

C:\<dev folder>\electronPOC\node_modules\winax>if not defined npm_config_node_gyp (node "C:\Program Files (x86)\nodejs\node_modules\npm\node_modules\npm-lifecycle\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "C:\Program Files (x86)\nodejs\node_modules\npm\node_modules\node-gyp\bin\node-gyp.js" rebuild )
Building the projects in this solution one at a time. To enable parallel build, please add the "-m" switch.
  main.cpp
  utils.cpp
  disp.cpp
  win_delay_load_hook.cc
C:\<my login>\AppData\Local\node-gyp\Cache\12.0.2\include\node\v8.h(1650,55): warning C4996: 'v8::Module::ResolveCal
lback': Use ResolveModuleCallback (compiling source file ..\src\utils.cpp) [C:\<dev folder>
\electronPOC\node_modules\winax\build\node_activex.vcxproj]
C:\<my login>\AppData\Local\node-gyp\Cache\12.0.2\include\node\v8.h(1650,55): warning C4996: 'v8::Module::ResolveCal
lback': Use ResolveModuleCallback (compiling source file ..\src\main.cpp) [C:\<dev folder>\
electronPOC\node_modules\winax\build\node_activex.vcxproj]
C:\<my login>\AppData\Local\node-gyp\Cache\12.0.2\include\node\v8.h(1650,55): warning C4996: 'v8::Module::ResolveCal
lback': Use ResolveModuleCallback (compiling source file ..\src\disp.cpp) [C:\<dev folder>\
electronPOC\node_modules\winax\build\node_activex.vcxproj]
     Creating library C:\<dev folder>\electronPOC\node_modules\winax\build\Release\node_act
  ivex.lib and object C:\<dev folder>\electronPOC\node_modules\winax\build\Release\node_act
  ivex.exp
disp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class std::shared_ptr<class v8::Bac
kingStore> __thiscall v8::ArrayBuffer::GetBackingStore(void)" (__imp_?GetBackingStore@ArrayBuffer@v8@@QAE?AV?$shared_pt
r@VBackingStore@v8@@@std@@XZ) referenced in function "private: static void __cdecl DispObject::NodeCreate(class v8::Fun
ctionCallbackInfo<class v8::Value> const &)" (?NodeCreate@DispObject@@CAXABV?$FunctionCallbackInfo@VValue@v8@@@v8@@@Z)
[C:\<dev folder>\electronPOC\node_modules\winax\build\node_activex.vcxproj]
C:\<dev folder>\electronPOC\node_modules\winax\build\Release\node_activex.node : fatal erro
r LNK1120: 1 unresolved externals [C:\<dev folder>\electronPOC\node_modules\winax\build\nod
e_activex.vcxproj]
gyp ERR! build error
gyp ERR! stack Error: `C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\MSBuild\Current\Bin\MSBuild.exe` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onExit (C:\Program Files (x86)\nodejs\node_modules\npm\node_modules\node-gyp\lib\build.js:194:23)
gyp ERR! stack     at ChildProcess.emit (events.js:315:20)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:277:12)
gyp ERR! System Windows_NT 10.0.19042
gyp ERR! command "C:\\Program Files (x86)\\nodejs\\node.exe" "C:\\Program Files (x86)\\nodejs\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd C:\<dev folder>\electronPOC\node_modules\winax
gyp ERR! node -v v14.16.0
gyp ERR! node-gyp -v v5.1.0
gyp ERR! not ok
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! winax@3.1.1 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the winax@3.1.1 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

**Comment by alx-dmt** [comment]
Hello. I had the same problem. Try npm install winax@1.20.0, I downloaded the old version directly - it worked.

**Comment by alx-dmt** [comment]
Hello. I had the same problem. Try npm install winax@1.20.0, I downloaded the old version directly - it worked.

**Comment by chpuget** [comment]
Hi, following the comment : "Hello. I had the same problem. Try npm install winax@1.20.0, I downloaded the old version directly - it worked." : thanks for advice !

I confirm it works well with the version 1.20.0. Tried this :
npm install winax@1.20.0 --msvs_version=2019
npm rebuild winax --runtime=electron --target=12.0.2 --dist-url=https://electronjs.org/headers --build-from-source

It would be perfect if it worked with the latest version

**Comment by durs** [comment]
try with last winax@3.1.3 version

**Comment by chpuget** [comment]
Hi durs, thanks for your quick answer,
Just tried and I see the same issue : 
npm install winax --msvs_version=2019
npm rebuild winax --runtime=electron --target=12.0.2 --dist-url=https://electronjs.org/headers --build-from-source

> winax@3.1.3 install C:\<dev folder>\electronPOC\node_modules\winax
> node-gyp rebuild


C:\<dev folder>\electronPOC\node_modules\winax>if not defined npm_config_node_gyp (node "C:\Program Files (x86)\nodejs\node_modules\npm\node_modules\npm-lifecycle\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "C:\Program Files (x86)\nodejs\node_modules\npm\node_modules\node-gyp\bin\node-gyp.js" rebuild )
Building the projects in this solution one at a time. To enable parallel build, please add the "-m" switch.
  main.cpp
  utils.cpp
  disp.cpp
  win_delay_load_hook.cc
C:\Users\<login>\AppData\Local\node-gyp\Cache\12.0.2\include\node\v8.h(1650,55): warning C4996: 'v8::Module::ResolveCal
lback': Use ResolveModuleCallback (compiling source file ..\src\utils.cpp) [C:\<dev folder>
\electronPOC\node_modules\winax\build\node_activex.vcxproj]
C:\Users\<login>\AppData\Local\node-gyp\Cache\12.0.2\include\node\v8.h(1650,55): warning C4996: 'v8::Module::ResolveCal
lback': Use ResolveModuleCallback (compiling source file ..\src\main.cpp) [C:\<dev folder>\
electronPOC\node_modules\winax\build\node_activex.vcxproj]
C:\Users\<login>\AppData\Local\node-gyp\Cache\12.0.2\include\node\v8.h(1650,55): warning C4996: 'v8::Module::ResolveCal
lback': Use ResolveModuleCallback (compiling source file ..\src\disp.cpp) [C:\<dev folder>\
electronPOC\node_modules\winax\build\node_activex.vcxproj]
     Creating library C:\<dev folder>\electronPOC\node_modules\winax\build\Release\node_act
  ivex.lib and object C:\<dev folder>\electronPOC\node_modules\winax\build\Release\node_act
  ivex.exp
disp.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class std::shared_ptr<class v8::Bac
kingStore> __thiscall v8::ArrayBuffer::GetBackingStore(void)" (__imp_?GetBackingStore@ArrayBuffer@v8@@QAE?AV?$shared_pt
r@VBackingStore@v8@@@std@@XZ) referenced in function "private: static void __cdecl DispObject::NodeCreate(class v8::Fun
ctionCallbackInfo<class v8::Value> const &)" (?NodeCreate@DispObject@@CAXABV?$FunctionCallbackInfo@VValue@v8@@@v8@@@Z)
[C:\<dev folder>\electronPOC\node_modules\winax\build\node_activex.vcxproj]
C:\<dev folder>\electronPOC\node_modules\winax\build\Release\node_activex.node : fatal erro
r LNK1120: 1 unresolved externals [C:\<dev folder>\electronPOC\node_modules\winax\build\nod
e_activex.vcxproj]

**Comment by durs** [comment]
Can't check VS2019 and node version12.0.2: 
`Error: 404 response downloading https://nodejs.org/dist/v12.0.2/node-v12.0.2-headers.tar.gz`

But VS2017 and node 12.22.0 (latest-v12.x) rebuild successfully 
```
npm rebuild winax --runtime=electron --target=12.22.0 --dist-url=https://electronjs.org/headers --build-from-source

> winax@3.1.3 install C:\test\node_modules\winax
> node-gyp rebuild

C:\test\node_modules\winax>if not defined npm_config_node_gyp (node "C:\nodejs\node_modules\npm\node_modules\npm-lifecycle\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "C:\nodejs\node_modules\npm\node_modules\node-gyp\bin\node-gyp.js" rebuild )
Building the projects in this solution one at a time. To enable parallel build, please add the "-m" switch.
  main.cpp
  utils.cpp
  disp.cpp
  win_delay_load_hook.cc
      C:\test\node_modules\winax\build\Release\node_activex.lib 
      C:\test\node_modules\winax\build\Release\node_activex.exp
     node_activex.vcxproj -> C:\test\node_modules\winax\build\Release\\node_activex.node
winax@3.1.3 C:\test\node_modules\winax

```

**Comment by chpuget** [comment]
Hi durs, I am using electron 12.0.2 because it's the recommended stable version, as suggested on electron.org :

npm i -D electron@latest --> this brings electron version 12.0.2

![image](https://user-images.githubusercontent.com/77572374/113781600-37c85680-9731-11eb-94ee-99e93fdce25a.png)

I have just retried with VS2017 instead of VS2019, but get the same error

Regards

**Comment by edo2313** [comment]
Hi. I have the same error. First getting a 404 with Electron 12.0.4, switched it manually to 12.0.0 seems to get rid of the 404.
`npm rebuild winax --runtime=electron --target=12.0.0 --dist-url=https://electronjs.org/headers --build-from-source` works fine, but when I start the app with `npx electron .` it complains about the different version (`npx electron -v` shows 12.0.0). Also when started with electron-forge, it can't compile and throws an LNK2001 error. Any ideas? I can test various things/combinations if you need.
(oh btw I have VS2019 installed)

**Comment by durs** [comment]
try new fix winax@3.1.4

**Comment by chpuget** [comment]
thanks a lot, closing the issue
